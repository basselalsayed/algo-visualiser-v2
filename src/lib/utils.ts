import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type T0 =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export function assert(value: unknown, type: T0): asserts value is T0 {
  const actualType = typeof value;

  if (type === 'object') {
    if (value === null) {
      throw new TypeError(`Expected type '${type}', but received null`);
    }
  }

  if (actualType !== type) {
    throw new TypeError(
      `Expected type '${type}', but received '${actualType}'`
    );
  }
}

export function typeIs(value: unknown, type: T0): value is T0 {
  try {
    assert(value, type);
    return true;
  } catch {
    return false;
  }
}

export function convertToMilliseconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const milliseconds = end - start;
  return Math.floor(milliseconds * 1000) / 1000; // 3 decimal places
}

export function convertToSeconds(
  start: DOMHighResTimeStamp,
  end: DOMHighResTimeStamp
) {
  const seconds = (end - start) / 1000;
  return Math.floor(seconds * 1000) / 1000;
}

export function elementIsFullyScrolled(element: HTMLElement): boolean {
  return (
    Math.floor(element.scrollHeight - element.scrollTop) ===
    Math.floor(element.clientHeight)
  );
}
