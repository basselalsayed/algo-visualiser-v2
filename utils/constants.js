import path from 'node:path';

export const DIRECTORIES = [
  'algorithms',
  'components',
  'contexts',
  'data',
  'hooks',
  'lib',
  'store',
];

export const ROOT_PATH = path.resolve(import.meta.dirname, '..');
