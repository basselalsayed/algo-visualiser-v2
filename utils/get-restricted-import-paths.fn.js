import fs from 'node:fs';
import path from 'node:path';

const __dirname = new URL('.', import.meta.url).pathname;

/**
 * Recursively walks through a directory and collects relative paths to all index.ts or index.tsx files.
 *
 * @param {string} dir - Absolute path to start walking from.
 * @param {string} baseRelativePath - Relative path from the module root (used for nested structure).
 * @returns {string[]} - List of relative paths to directories containing index files.
 */
function findIndexFileDirs(dir, baseRelativePath = '') {
  const indexDirs = [];
  const entries = fs.readdirSync(dir);

  for (const entry of entries) {
    const fullPath = path.join(dir, entry);
    const relPath = path.join(baseRelativePath, entry);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      indexDirs.push(...findIndexFileDirs(fullPath, relPath));
    } else if (
      (entry === 'index.ts' || entry === 'index.tsx') &&
      baseRelativePath !== ''
    ) {
      indexDirs.push(baseRelativePath);
    }
  }

  return indexDirs;
}

/**
 * Generates a list of restricted import paths for all index.ts(x) files in a module.
 *
 * @param {string} dirAbsolutePath - Absolute path to the module directory (e.g. src/components)
 * @param {string} alias - Import alias for the module (e.g. @/components)
 * @returns {Array<{ name: string, message: string }>}
 */
function getRestrictedIndexImports(dirAbsolutePath, alias) {
  const indexDirs = findIndexFileDirs(dirAbsolutePath);

  const staticUpwardPaths = [
    '../../../index',
    '../../..',
    '../../index',
    '../..',
    '../index',
    '..',
  ].map((name) => ({
    message: `Avoid importing '${name}'. Use aliased or direct file paths.`,
    name,
  }));

  const indexPathViolations = indexDirs.flatMap((relPath) => [
    {
      message: `Avoid importing index file: ${alias}/${relPath}/index.ts. Import specific files instead.`,
      name: `${alias}/${relPath}`,
    },
    {
      message: `Avoid importing index file: ../${relPath}/index.ts. Import specific files instead.`,
      name: `../${relPath}`,
    },
  ]);

  return [...staticUpwardPaths, ...indexPathViolations];
}

export { __dirname, getRestrictedIndexImports };
