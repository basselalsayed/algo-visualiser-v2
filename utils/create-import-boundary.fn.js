import path from 'node:path';

import { DIRECTORIES, ROOT_PATH } from './constants.js';
import { getRestrictedIndexImports } from './get-restricted-import-paths.fn.js';

/**
 * @param {string} dir
 */
const getAlias = (dir) => `@/${dir}`;

/**
 * @param {string} dir
 */
export function createImportBoundary(dir) {
  const alias = getAlias(dir);
  const glob = `src/${dir}/**/*.{ts,tsx}`;

  const forbid = DIRECTORIES.flatMap((_dir) =>
    _dir === dir ? undefined : [`${getAlias(_dir)}/**/*`, `${getAlias(_dir)}/*`]
  ).filter(Boolean);

  const absoluteDir = path.resolve(ROOT_PATH, 'src', dir);

  const restrictedIndexPaths = getRestrictedIndexImports(absoluteDir, alias);

  return {
    files: [glob],
    rules: {
      'import/no-internal-modules': [
        'error',
        {
          forbid,
        },
      ],
      'no-restricted-imports': [
        'error',
        {
          paths: [
            ...restrictedIndexPaths,
            {
              message: `Avoid importing '${alias}' from within ${dir}/. Use direct file paths.`,
              name: alias,
            },
          ],
          patterns: [
            {
              group: ['**/index', '**/index.ts', '**/index.tsx'],
              message: `Avoid importing index files directly. Import specific files instead.`,
            },
            {
              group: ['../../**/*'],
              message: `Avoid relative imports outside of this module. Use alias paths instead.`,
            },
          ],
        },
      ],
    },
  };
}
