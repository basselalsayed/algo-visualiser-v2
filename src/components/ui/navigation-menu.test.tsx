import { renderFragmentSnapshot } from '@/test/utils';

import {
  NavigationMenu,
  NavigationMenuContent,
  NavigationMenuItem,
  NavigationMenuList,
  NavigationMenuTrigger,
} from './navigation-menu';

describe('NavigationMenu.component', () => {
  it('looks the same', () => {
    expect(
      renderFragmentSnapshot(
        <NavigationMenu>
          <NavigationMenuList>
            <NavigationMenuItem>
              <NavigationMenuTrigger />
              <NavigationMenuContent />
            </NavigationMenuItem>
          </NavigationMenuList>
        </NavigationMenu>
      )
    ).toMatchSnapshot();
  });
});
