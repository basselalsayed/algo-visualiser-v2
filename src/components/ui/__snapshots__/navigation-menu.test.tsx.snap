// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationMenu.component > looks the same 1`] = `
<DocumentFragment>
  <nav
    aria-label="Main"
    class="relative z-10 flex max-w-max flex-1 items-center justify-center"
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      style="position: relative;"
    >
      <ul
        class="group flex flex-1 list-none items-center justify-center space-x-1"
        data-orientation="horizontal"
        dir="ltr"
      >
        <li>
          <button
            aria-controls="radix-«r0»-content-radix-«r1»"
            aria-expanded="false"
            class="gap-2 whitespace-nowrap focus-visible:ring-1 focus-visible:ring-ring focus-visible:outline-hidden [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 animate_grad bg_grad_accent--outline--text shadow-sm [--base-color:hsl(var(--background))] group inline-flex h-9 w-max items-center justify-center rounded-md bg-background px-4 py-2 text-sm font-medium transition-colors hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground focus:outline-hidden disabled:pointer-events-none disabled:opacity-50 data-active:bg-accent/50 data-[state=open]:bg-accent/50 group"
            data-radix-collection-item=""
            data-state="closed"
            id="radix-«r0»-trigger-radix-«r1»"
          >
            <svg
              aria-hidden="true"
              class="relative h-3 w-3 transition duration-300 group-data-[state=open]:rotate-180"
              fill="none"
              height="15"
              viewBox="0 0 15 15"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.13523 6.15803C3.3241 5.95657 3.64052 5.94637 3.84197 6.13523L7.5 9.56464L11.158 6.13523C11.3595 5.94637 11.6759 5.95657 11.8648 6.15803C12.0536 6.35949 12.0434 6.67591 11.842 6.86477L7.84197 10.6148C7.64964 10.7951 7.35036 10.7951 7.15803 10.6148L3.15803 6.86477C2.95657 6.67591 2.94637 6.35949 3.13523 6.15803Z"
                fill="currentColor"
                fill-rule="evenodd"
                stroke="url(#gradient-sVnBjCeMY2eSwZoLjIe9Z)"
              />
              <defs>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="gradient-sVnBjCeMY2eSwZoLjIe9Z"
                  style="--grad-1: hsl(var(--grad-accent-1)); --grad-2: hsl(var(--grad-accent-2));"
                  x1="0%"
                  x2="100%"
                  y1="0%"
                  y2="100%"
                >
                  <stop
                    offset="0%"
                    style="stop-color: var(--grad-1);"
                  />
                  <stop
                    offset="100%"
                    style="stop-color: var(--grad-2);"
                  />
                </lineargradient>
              </defs>
            </svg>
          </button>
        </li>
      </ul>
    </div>
    <div
      class="absolute top-full right-0 flex justify-center"
    />
  </nav>
</DocumentFragment>
`;
